### PR TITLE
fix: Fix calculation of AABB for `ShapeHitbox`es

### LINKED ISSUE #1478: Collision callbacks start and end are called in the same tick with many added hitboxes
# Current bug behaviour
When two hitboxs has same size and overlaps perfectly,
`onCollisionStart` and `onCollisionEnd` callback called alternately and indefinitely even though there is no position changed. 

# Expected behaviour
I don't know actually but at least current behavior is weired...

# Steps to reproduce
1. Create Two components that has hitbox
2. Set hitbox size as same as each other

# Flutter doctor output
```
[✓] Flutter (Channel stable, 2.10.3, on macOS 12.1 21C52 darwin-arm, locale ko-KR)
    • Flutter version 2.10.3 at /Users/lafity101/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (3 weeks ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/lafity101/Library/Android/sdk
    • Platform android-32, build-tools 32.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.65.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (2 available)
    • iPhone (mobile) • DEVICE_ID • ios            • iOS 15.1 19B74
    • Chrome (web)          • chrome                    • web-javascript • Google Chrome 99.0.4844.74

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

# More environment information

* Flame version: `main` branch

# Log information

```dart
[ +31 ms] flutter: start collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ ] flutter: end collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ +33 ms] flutter: start collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ ] flutter: end collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ +31 ms] flutter: start collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ ] flutter: end collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ +16 ms] flutter: start collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
[ ] flutter: end collision other: [2448.0,1296.0], size: [32.0,32.0] // me: center: [2448.0,1296.0], size: [32.0,52.0]
... you can see this at the following video
```

https://user-images.githubusercontent.com/38072762/159412816-5b081754-77a1-40f5-aee3-76358794aaab.mov

### PR DESCRIPTION
**Pull request by spydon**
# Description

Tried to reproduce #1478 without luck.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

Closes #1478 

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/test/collision_detection/collision_callback_test.dart [hunk 1]
```diff
@@ -198,4 +198,48 @@ void main() {
...
+      );
+      await game.ensureAddAll([blockA, blockB]);
+      game.update(0);
+      expect(blockA.startCounter, 1);
+      expect(blockB.startCounter, 1);
+      expect(blockA.onCollisionCounter, 1);
+      expect(blockB.onCollisionCounter, 1);
+      expect(blockA.endCounter, 0);
+      expect(blockB.endCounter, 0);
+      game.update(0);
```
Comment (erickzanardo) [inline packages/flame/test/collision_detection/collision_callback_test.dart:221]: Nit: add a blank line between the expect groups to make reading easier

Comment (spydon) [reply packages/flame/test/collision_detection/collision_callback_test.dart:221]: They are quite clearly separated by the `game.update` I think?
Or did I misunderstand where you wanted a new line?

Comment (erickzanardo) [reply packages/flame/test/collision_detection/collision_callback_test.dart:221]: I mean like this:

```dart
      game.update(0);
      expect(blockA.startCounter, 1);
      expect(blockB.startCounter, 1);
      expect(blockA.onCollisionCounter, 3);
      expect(blockB.onCollisionCounter, 3);
      expect(blockA.endCounter, 0);
      expect(blockB.endCounter, 0);
      blockB.position.y += 20;

      game.update(0);
      expect(blockA.startCounter, 1);
      expect(blockB.startCounter, 1);
      expect(blockA.onCollisionCounter, 3);
      expect(blockB.onCollisionCounter, 3);
      expect(blockA.endCounter, 1);
      expect(blockB.endCounter, 1);
```

Comment (spydon) [reply packages/flame/test/collision_detection/collision_callback_test.dart:221]: Might visually look better, but I think that the position belongs to the next block at least, but maybe that was just a typo?

**Comment by st-pasha** [comment]
What would be the reported set of intersection points in this case?

**Comment by spydon** [comment]
> I mean, normally when 2 shapes collide, we compute their set of collision points. But here the two shapes overlap exactly, so I was wondering what's going to happen to that set (in theory there are infinite number of collision points now). A circle-on-circle overlap would be interesting too...

Aah, now I get what you mean, with polygon collisions that have lines that are on each other the intersections points will be formed by the line that is created by each overlap and the points will be the start, end and middle of that line.
For circles that are exactly on top of each other four intersection points with pi/2 angle between them will be reported.

File: packages/flame/test/collision_detection/collision_callback_test.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/test/collision_detection/collision_callback_test.dart:232]: Split them up like this, it was actually much better.

**Review by erickzanardo — APPROVED** [review]

**Review by luanpotter — APPROVED** [review]

**Pull request merged**